geo/src/algorithm/centroid.rs

```diff
@@ -340,28 +339,153 @@ where
     type Output = Option<Point<T>>;
 
     fn centroid(&self) -> Self::Output {
-        if self.0.is_empty() {
-            return None;
+        centroid_of_coords(self.iter().map(|p| p.0)).map(Point)
+    }
+}
+
+impl<T> Centroid for Geometry<T>
+where
+    T: CoordFloat + FromPrimitive + Sum + Default,
+{
+    type Output = Option<Point<T>>;
+
+    crate::geometry_delegate_impl! {
+        fn centroid(&self) -> Self::Output;
+    }
+}
+
+impl<T> Centroid for GeometryCollection<T>
+where
+    T: CoordFloat + FromPrimitive + Sum + Default,
+{
+    type Output = Option<Point<T>>;
+    fn centroid(&self) -> Self::Output {
+        use crate::algorithm::dimensions::{Dimensions, HasDimensions};
+
+        // The Geometries in the GeometryCollection could have different dimensionality. Centroids
+        // must be considered separately by dimensionality.
+        //
+        // e.g. If I have several Points, adding a new `Point` will affect their centroid.
+        //
+        // However, because a Point is zero dimensional, it is infinitely small when compared to
+        // any 2-D Polygon. Thus a Point will not affect the centroid of any GeometryCollection
+        // containing a 2-D Polygon.
+        //
+        // So, for each Geometry in the GeometryCollection, we accumulate that Geometry's
+        // `(centroid, weight)` with that of other geometries of the same dimensionality. And at
+        // the end, the centroid of the GeometryCollection as a whole is the highest dimensional
+        // centroid which has `Some` value.
+
+        // (accumulated centroid, accumulated weight) tuples for each dimensionality.
+        let mut centroid_0d_accum: Option<(Point<T>, T)> = None;
+        let mut centroid_1d_accum: Option<(Point<T>, T)> = None;
+        let mut centroid_2d_accum: Option<(Point<T>, T)> = None;
+
+        for geometry in self.iter() {
+            let centroid = geometry.centroid();
+            match (geometry.dimensions(), centroid) {
+                (Dimensions::Empty, _) | (_, None) => continue,
+                (Dimensions::ZeroDimensional, Some(centroid)) => {
+                    let mut centroid_accum = centroid_0d_accum.unwrap_or_default();
+                    let weight = T::one();
```

Comment:
ooph good catch. I'm going to rework this, but might not have for a few days. 

Thank you so much for catching this mistake!

Reply:
Ok, I believe I've addressed this, but it ended being non-trivial. In the process I found a few more bugs and fleshed out some more tests (almost always the issues were with degenerate geometries).

> I think we should extend our Centroid trait to add a method that returns both the centroid Point, and the total weight of the geometry.

I started with this approach, and it would've worked fine, but I felt things ended up cleaner when I bundled the state and logic into a separate private `CentroidOperation` accumulation struct.

Even in cases where there is a single geometry, we'll still utilize the aggregating operation, which at first seems weird.  This is primarily to avoid redundantly having to recompute the centroids weight in the case the geometry *does* appear in a collection. Since computing the centroid essentially involves computing its weight anyway, there's not much lost by tracking it in this way, except some small extra stack allocations in some cases.

Since I've essentially re-written the trait at this point, it'd be great if I could get another review.